# Python - Einstieg - Teil 1 - Datentypen

Themen:
* Übersicht Datentypen in Python:
* Datentyp Zahlen - Integer und Fließkommazahlen

2026-03-25 ug Version 1.3

<!---
Todo:

--->

### Referenz:

Die [_Python Language Reference_](https://docs.python.org/3/reference/index.html#reference-index) beschreibt offiziell die Programmiersprache **_Python_**.

# Übersicht der Datentypen in Python (unvollständig):

Wir betrachten hier folgende 
einfache Datentypen: 
* Zahlen - Integer und Fließkommazahlen
* Strings - Zeichenketten
* Booleans - True oder False

und folgende höhere Datentypen:
* Lists, Tuples - geordnete Speicherung von Datenelementen
* Dictionaries - Schlüssel-Werte-Paare 
* Sets - "Mengen" - ungeordnete Ansammlung von Werte ohne Doppelungen

Die Aufstellung ist nicht vollständig, es gibt noch weitere Datentypen siehe [ The Python Standard Library » Built-in Types](https://docs.python.org/3/library/stdtypes.html).

-------
# Datentypen Zahlen - Integer und Fließkommazahlen

## Python kann als "Taschenrechner" eingesetzt werden

### Grundrechenarten: Addition, Subtraktion, Multiplikation, Division

In [1]:
1+3

4

In [2]:
1-3

-2

In [3]:
3*4

12

In [4]:
5/2

2.5

Python unterscheidet zwischen folgenden Zahlentypen:
   - Ganzzahlen (Integer) -> Zahlen ohne Dezimaltrenner `.` z.B.: 5
   - Fließkommazahlen (Floatingpointzahlen):
       - Zahlen mit Dezimaltrenner `.`  z.B.: 3.1415 oder  3.0
       - Zahlen mit Exponentschreibweise z.B.: 1e10
   
Die Addition (`+`), die Subtraktion (`-`) und die Multiplikation (`*`) von zwei Integerzahlen, liefert wieder eine Integerzahl.

Die Division (`/`) von zwei Integern liefert immer eine Fließkomma-Zahl.

Mit der [type()](https://www.geeksforgeeks.org/python-type-function/)-Funktion wird der "Typ" eines Python Objekts also auch der einer Zahl ([Literal](https://de.wikipedia.org/wiki/Literal)) bestimmt.

- `int`: Integer - Ganzzahl
- `float`: Fließkommanzahl

Ausblick: Die `type`-Funktion wird bei _Mutable und Immutable Objects_ wieder aufgegriffen.


In [5]:
type(3)

int

In [6]:
type(2.5)

float

In [7]:
type(1e10)

float

In [8]:
type(1+3)

int

In [9]:
type(5/2)

float

In [10]:
4/2

2.0

In [11]:
type(4/2)

float

In [12]:
int(1e10)

10000000000

In [13]:
float(2)

2.0

Zur besseren Lesbarkeit von langen Zahlen können Underscores `_` an beliebiger Stelle eingefügt werden. 

Python ignoriert beim Lesen der Zahl die Underscores, siehe [PEP 515 – Underscores in Numeric Literals](https://peps.python.org/pep-0515/).

In [14]:
10_000_000_000

10000000000

In [15]:
10_00_0000_000

10000000000

In [16]:
type(10_000_000_000)

int

In [17]:
float(10_000_000_000)

10000000000.0

Darstellung einer Zahl in der Exponentenschreibweise. Hierzu werden f-Strings (formatierte Zeichenketten) verwenden, die wir später noch ausführlich behandeln.

In [18]:
f"{10_000_000_000:1.1e}"

'1.0e+10'

## ganzzahlige Division und Division mit Rest (modulo)

- [floor division](https://python-reference.readthedocs.io/en/latest/docs/operators/floor_division.html) `//` ganzzahlige Division mit "Abrundungsfunktion"
- [modulus](https://python-reference.readthedocs.io/en/latest/docs/operators/modulus.html) `%` "Rest"

Es gilt folgender Zusammenhang zwischen Quotient, Dividend, Divisior, Zähler und Nenner; vergleiche [Division - Wikipedia](https://de.wikipedia.org/wiki/Division_(Mathematik)):
$$
    Quotient = \frac{\text{Dividend}}{\text{Divisor}} = \frac{\text{Zähler}}{\text{Nenner}}
$$

Die "_Floor division_" `//` (ganzzahlige Division mit "Abrundungsfunktion") ist definiert als:
$$
    \text{Zähler} ~ \text{//} ~ \text{Nenner} = \text{floor}(\text{Zähler} ~ \text{/} ~ \text{Nenner})
$$
mit $\text{floor}()$  als der Abrundungsfunktion.

Die Abrundungsfunktion  $\text{floor}(x)$ liefert die größte Integer, die kleiner oder gleich der Zahl x im Argument ist:
- $\text{floor}()$ für positive Argumente: $\text{floor}(3.2)$ = 3 (dies entspricht dem Abschneiden der Nachkommanstellen)
- $\text{floor}()$ für negative Argumente: $\text{floor}(-3.2)$ = -4 (abrunden heißt in Richtung minus unendlich gehen)

Die Division mit Rest, vergleiche [Division mit Rest - Wikipedia](https://de.wikipedia.org/wiki/Division_mit_Rest) oder
[Python Floor Division - pythontutorial.net](https://www.pythontutorial.net/advanced-python/python-floor-division) wird durchfolgende Gleichung beschrieben:

$$
\begin{align}
    \text{Zähler} &= \text{Nenner} ~ * ~ \underbrace{(\text{Zähler} ~ \text{//} ~ \text{Nenner})}_{\text{floor division}} ~+~ \underbrace{(\text{Zähler}~\text{\%}~\text{Nenner})}_{\text{modolo}} \\
    \text{Z} &= \text{N} ~ * ~\underbrace{(\text{Z} ~ \text{//} ~ \text{N})}_{\text{floor division}}~+~ \underbrace{(\text{Z}~\text{\%}~\text{N})}_{\text{modolo}}   
\end{align}
$$
mit $\text{Z}$ für Zähler und $\text{N}$ für Nenner.

Das Ergebnis der Division ist der ganzzahlige Anteil der Division. Falls beide Operanden Integerwerte sind, stellt auch das Ergebnis einen Integerwert dar. Falls jedoch mindestens einer der Operanden ein Floatwert ist, so wird der ganzzahlige Anteil der Division als Float ausgegeben. 



Referenzen:
- [Division - Wikipedia](https://de.wikipedia.org/wiki/Division_(Mathematik))
- [Division mit Rest - Wikipedia](https://de.wikipedia.org/wiki/Division_mit_Rest)
- [Python3 Operatoren - python-kurs.eu](https://www.python-kurs.eu/python3_operatoren.php)
- [Python Floor Division - pythontutorial.net](https://www.pythontutorial.net/advanced-python/python-floor-division)
- [Python Modulo - pythontutorial.net](https://www.pythontutorial.net/advanced-python/python-modulo)
- [Abrundungsfunktion_und_Aufrundungsfunktion - Wikipedia](https://de.wikipedia.org/wiki/Abrundungsfunktion_und_Aufrundungsfunktion)
    

**Beispiel - nur positive Zahlen**: 5 durch 2 ist 2 Rest 1

In [19]:
5//2

2

In [20]:
type(5//2)

int

In [21]:
5%2 

1

In [22]:
type(5%2)

int

Was passiert bei negativen Zahlen?

**Beispiel - mit einer negativen Zahl**: -5 durch 2 ist -3 Rest 1

In [23]:
-5//2

-3

In [24]:
-5%2

1

Nebenrechnung zu "Beispiel - mit einer negativen Zahl":

Es gilt die Formel:
$$
    \text{Z} = \text{N} ~ * ~ (\text{Z} ~ \text{//} ~ \text{N}) ~+~ (\text{Z}~\text{\%}~\text{N})   
$$
mit $\text{Z}=-5$ und $\text{N}=2$.

Die float Division ist:
$$
    \text{Z} ~ \text{/} ~ \text{N} = -5  ~ \text{/} ~ 2= -2.5 
$$
Abgerunden ergibt dies -3; dies ist die floor division:
$$
    \text{Z} ~ \text{//} ~ \text{N} = \text{-5} ~ \text{//} ~ \text{2} = -3
$$

Dann ergibt sich durch Umstellen der oben genannten Formel für die Modulo Operation $(\text{Z}~\text{\%}~\text{N})$:
$$
\begin{align}
    \text{Z}~\text{\%}~\text{N}  &= \text{Z}- \text{N} ~ * ~ (\text{Z} ~ \text{//} ~ \text{N}) \\
    \text{Z}~\text{\%}~\text{N}  &= (-5) - (2) ~ * ~ (-3) \\
    \text{Z}~\text{\%}~\text{N}  &= (-5) - (-6) \\
    \text{Z}~\text{\%}~\text{N}  &= 1   
\end{align}
$$

------
#### Was liefert die floor division bei realen Zahlen (float)?

Das Ergebnis ist eine Float-Zahl, sobald einer der Operanden eine Float-Zahl ist.

In [25]:
5//2.0

2.0

In [26]:
type(5//2.0)

float

In [27]:
5.0//2

2.0

In [28]:
type(5.0//2)

float

Was passiert wenn wir keine ganzen Zahlen verwenden?

In [29]:
5//2.1

2.0

In [30]:
5%2.1

0.7999999999999998

In [31]:
print(">0123456789<")
print(f">{5%2.1:10.2f}<")

>0123456789<
>      0.80<


In [32]:
5/2.1

2.380952380952381

Es gilt die Formel:
$$
    \text{Z} = \text{N} ~ * ~ (\text{Z} ~ \text{//} ~ \text{N}) ~+~ (\text{Z}~\text{\%}~\text{N})   
$$
mit $\text{Z}=5$ und $\text{N}=2.1$.

Die float Division ist:
$$
    \text{Z} ~ \text{/} ~ \text{N} = 5  ~ \text{/} ~ 2.1= 2.380952380952381
$$
Abgerunden ergibt dies -3; dies ist die floor division:
$$
    \text{Z} ~ \text{//} ~ \text{N} = \text{5} ~ \text{//} ~ \text{2.1} = 2
$$

Dann ergibt sich durch Umstellen der oben genannten Formel für die Modulo Operation $(\text{Z}~\text{\%}~\text{N})$:
$$
\begin{align}
    \text{Z}~\text{\%}~\text{N}  &= \text{Z}- \text{N} ~ * ~ (\text{Z} ~ \text{//} ~ \text{N}) \\
    \text{Z}~\text{\%}~\text{N}  &= (5) - (2.1) ~ * ~ (2) \\
    \text{Z}~\text{\%}~\text{N}  &= (5) - (4.2) \\
    \text{Z}~\text{\%}~\text{N}  &= 0.8   
\end{align}
$$

Beim Rechnen mit Fließkommazahlen entstehen Rundungsfehler.

------
## Potenzieren

Potenzieren: Basis `**` Exponent

In [33]:
2**8

256

In [34]:
type(2**8)

int

**Anmerkung:**

Das Potenzieren (`**`) von zwei Integerzahlen, liefert wieder eine Integerzahl.


Das Ergebnis ist eine Float-Zahl, sobald einer der Operanden eine Float-Zahl ist.

In [35]:
type(2.0**8)

float

In [36]:
type(2**8.0)

float

----
## "Punkt vor Strichrechnung"

Bei der Bearbeitung eines Ausdruck gilt "Punkt vor Strickrechnung".
Multiplikation und Division wird vor Addition und Subtraktion ausgewertet.
Durch runde Klammern `()` kann die Reihenfolge aber eindeutig beschrieben werden.

In [37]:
2+3*4+5

19

In [38]:
(2+3)*(4+5)

45